Adding Fluid Dynamics to RomApp

Rbravo555 opened the pull request:
This PR allows the RomApplication to run reduced order models of fluid dynamics cases using the monolithic solver of the FluidDynamicsApplication

The test files retrieve the geometry directly from the CouetteFlowTest of the FluidDynamicsApplication

Rbravo555 committed:
erasing __pycache__ folder after tests

Rbravo555 committed:
Adding CFD Monolithic to ROMApp

Rbravo555 committed:
Adding test for CFD ROM (re-using geometry from FluidDynamicsApp's CouetteFlowTest test)

rubenzorrilla commented on applications/RomApplication/python_scripts/fluid_dynamics_analysis_rom.py:
```diff
@@ -0,0 +1,44 @@
+    def _GetSimulationName(self):
+        return "::[ROM Simulation]:: "
+
+    def ModifyInitialGeometry(self):
```
It is a bit counterintuitive to me to do this in the `ModifyInitialGeometry` function. Could it be done in `ModifyAfterSolverInitialize`? (not sure if you need your solver to have this values set before calling the `Initialize`. If this is the case, I think we could propose introducing a ` ModifyBeforeSolverInitialize` to the `AnalysisStage`.

Rbravo555 replied:
I have no objection. I require the nodes to have their ROM_BASIS and the Elements to have their ROM_WEIGHT before the SetSetUpDofSet method of the RomBuilder&Solver is called. I changed it as suggested

rubenzorrilla replied:
I think this is a better usage of the `AnalysisStage` methods. Thanks!

rubenzorrilla commented on applications/RomApplication/python_scripts/fluid_dynamics_analysis_rom.py:
```diff
@@ -0,0 +1,44 @@
+            for node in computing_model_part.Nodes:
+                aux = KratosMultiphysics.Matrix(nodal_dofs, rom_dofs)
+                for j in range(nodal_dofs):
+                    Counter=str(node.Id)
```
Why casting to string?

Rbravo555 replied:
We are calling the key in the JSON file in which the nodal modes are contained, that's why it is a string...

rubenzorrilla commented on applications/RomApplication/tests/fluid_dynamics_test_files/MainKratosROM.py:
```diff
@@ -0,0 +1,40 @@
+import sys
+import time
+import numpy as np
+import pdb
```
I think you're not using this

Rbravo555 replied:
addressed

rubenzorrilla commented on applications/RomApplication/tests/fluid_dynamics_test_files/test_ROM.py:
```diff
@@ -0,0 +1,39 @@
+class ROMFluidDynamics(KratosUnittest.TestCase):
+#########################################################################################
+
+    def test_Fluid_Dynamic_ROM_2D(self):
```
```suggestion
    def test_Fluid_Dynamics_ROM_2D(self):
```

Rbravo555 replied:
changed

rubenzorrilla commented on applications/RomApplication/tests/fluid_dynamics_test_files/test_ROM.py:
```diff
@@ -0,0 +1,39 @@
+                UP = sum((ExpectedOutput[:,i] - ObtainedOutput[:,i])**2)
+                DOWN = sum((ExpectedOutput[:,i])**2)
+                L2 = np.sqrt(UP/DOWN)
+                self.assertLess(L2, 1e-12)
```
```suggestion
                up=0
                down=0
                up = sum((ExpectedOutput[:,i] - ObtainedOutput[:,i])**2)
                down = sum((ExpectedOutput[:,i])**2)
                l2 = np.sqrt(up/down)
                self.assertLess(l2, 1e-12)
```

rubenzorrilla replied:
Besides the style guide issue, I think you don't need to initialize up and down to zero.

Rbravo555 replied:
changed

rubenzorrilla commented on applications/RomApplication/tests/test_RomApplication.py:
```diff
@@ -37,6 +38,7 @@ def AssembleTestSuites():
     smallSuite.addTest(ROMStaticStruct('test_Struct_Static_ROM_2D'))
     smallSuite.addTest(ROMStaticStruct('test_Struct_Static_HROM_2D'))
     smallSuite.addTest(ROMDynamicStruct('test_Struct_Dynamic_ROM_2D'))
+    smallSuite.addTest(ROMFluidDynamics('test_Fluid_Dynamic_ROM_2D'))
```
```suggestion
    smallSuite.addTest(ROMFluidDynamics('test_Fluid_Dynamics_ROM_2D'))
```

Rbravo555 replied:
changed

Rbravo555 committed:
Updating format of files

Rbravo555 committed:
Erasing unused variable AUX_ID

Rbravo555 committed:
Changing the method where the ROM_BASIS is imported to the nodes. From ModifyInitialGeometry to ModifyAfterSolverInitialize

Rbravo555 committed:
addressing PR comments

rubenzorrilla commented on applications/RomApplication/tests/structural_static_test_files/MainKratosHROM.py:
```diff
@@ -8,10 +6,10 @@
 
     def ModifyInitialGeometry(self):
-        super(TestStructuralMechanicsStaticHROM,self).ModifyInitialGeometry()
+        super().ModifyInitialGeometry()
```
Shouldn't this be `ModifyAfterSolverInitialize` after what we spoke before?

Rbravo555 replied:
Indeed, I have changed it as well now

Rbravo555 committed:
Imposing Elemental Weights in ModifyAfterSolverInitialize method, rather than ModifyInitialGeometry

rubenzorrilla reviewed (approved):
For me this is ready to go in.
